python/az/aro/azext_aro/custom.py

```diff
@@ -134,9 +136,45 @@ def aro_create(cmd,  # pylint: disable=too-many-locals
                        parameters=oc)
 
 
-def aro_delete(client, resource_group_name, resource_name, no_wait=False):
+def aro_delete(cmd, client, resource_group_name, resource_name, no_wait=False):
     # TODO: clean up rbac
 
+    oc = None
+    try:
+        oc = client.get(resource_group_name, resource_name)
+    except CloudError:
+        pass
+
+    if oc:
+        master_subnet = oc.master_profile.subnet_id
+        worker_subnets = [w.subnet_id for w in oc.worker_profiles]
+
+        master_parts = parse_resource_id(master_subnet)
+        vnet = resource_id(
+            subscription=master_parts['subscription'],
+            resource_group=master_parts['resource_group'],
+            namespace='Microsoft.Network',
+            type='virtualNetworks',
+            name=master_parts['name'],
+        )
+
+        aad = AADManager(cmd.cli_ctx)
+
+        client_id = oc.service_principal_profile.client_id
+
+        client_sp = aad.get_service_principal(client_id)
+        if not client_sp:
+            client_sp = aad.create_service_principal(client_id)
```

Comment:
Yikes, no!  If client_sp is None, this script can't recover it and we should just bottle out.
Same for rp_client_sp.

Reply:
What should the error message indicate, then? They will need a service principal to delete the cluster, no?

I don't want to just throw an error and have them file a support ticket that deletion failed, want to unblock the deletion.

Reply:
*I might be wrong here in my evaluations, so lets see if this make sense* 

I think deletion is happening using `fpAuthorizer`[1] so this implies we deleting resources from the customer RG. 

The issue here is that `client_sp` might not exist anymore. In the `client_sp` case this is challenging, as we do not support the update of `client_sp` in RP, so I think we don't have the right to fail and leave the customer hanging. My take on this is that we proceed with deletion but print out a warning stating that cluster SP was not found, and potentially was deleted. So we will not be able to clean RBAC for <insert client_SP_ID> for objects <insert_vnet_routetable_IDS>.

for `rp_client_sp` customers can't delete it, so we should be able to create assignments IF the customer has "User Access Administrator". Otherwise - error. 

1. https://github.com/Azure/ARO-RP/blob/175b63993a7085f945337297c3301e7472dae4a6/pkg/backend/openshiftcluster/openshiftcluster.go#L97